File: java/Trie.java
class TrieNode {
    TrieNode[] child = new TrieNode[26]; 
    String word; 
}

public class Trie {
    public TrieNode root = new TrieNode();

    public void insert(String word) {
        TrieNode node = root;
        for (var ch : word.toCharArray()) {
            int c = ch - 'a';
            if (node.child[c] == null)
                node.child[c] = new TrieNode();
            node = node.child[c];
        }
        node.word = word;
    }

    public boolean prefix(String word) {
        TrieNode node = root;
        for (var ch : word.toCharArray()) {
            int c = ch - 'a';
            if (node.child[c] == null) return false;
            node = node.child[c];
        }
        return true;
    }

    public boolean search(String word) {
        TrieNode node = root;
        for (var ch : word.toCharArray()) {
            int c = ch - 'a';
            if (node.child[c] == null) return false;
            node = node.child[c];
        }
        return node.word != null;
    }
}

File: cpp/trie.cpp
#include <string>

using namespace std;

struct TrieNode {
    TrieNode* child[26] = {nullptr};
    string word;
    ~TrieNode() { for (auto& node : child) delete node; }
};

class Trie {
public: 
    TrieNode* root = nullptr; 

    Trie() { root = new TrieNode(); }

    ~Trie() { delete root; }

    void insert(string word) {
        TrieNode* node = root; 
        for (auto& ch : word) {
            int c = ch - 'a';
            if (!node->child[c]) node->child[c] = new TrieNode();
            node = node->child[c];
        }
        node->word = word;
    }

    bool prefix(string word) {
        TrieNode* node = root; 
        for (auto& ch : word) {
            int c = ch - 'a';
            if (!node->child[c]) return false;
            node = node->child[c];
        }
        return true; 
    }

    bool search(string word) {
        TrieNode* node = root; 
        for (auto& ch : word) {
            int c = ch - 'a';
            if (!node->child[c]) return false;
            node = node->child[c];
        }
        return node->word != "";
    }
};
